Feature/kinesis sink

abhinavbansal246 opened the pull request:
Adding functionality to connect Hydrator realtime pipelines to Amazon Kinesis sink

russorat commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,136 @@
+      String state = streamState(kinesisClient, streamName);
+      switch (state) {
+        case "DELETING":
+          long waitTime = System.currentTimeMillis() + 1000 * 120;
```
maybe move the 1000*120 to a TIMEOUT variable.

russorat commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,136 @@
+          while (System.currentTimeMillis() < waitTime && streamExists(kinesisClient, streamName)) {
+            try {
+              LOG.info("...Deleting Stream {} ...", streamName);
+              Thread.sleep(1000 * 10);
```
and a SLEEP_INTERVAL variable

russorat commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,136 @@
+            }
+          }
+          if (streamExists(kinesisClient, streamName)) {
+            LOG.error("KinesisUtils timed out waiting for stream {} to delete", streamName);
```
if you are throwing the error, do you also need to log this?

russorat commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,136 @@
+          LOG.info("Stream {} is ACTIVE", streamName);
+          return;
+        case "CREATING":
+          break;
```
You log everywhere else but here?

russorat commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,136 @@
+      kinesisClient.createStream(createStreamRequest);
+      LOG.info("Stream {} created", streamName);
+    }
+    long waitTime = System.currentTimeMillis() + (10 * 60 * 1000);
```
a different wait time than before?

abhinavbansal246 replied:
Creation takes more time than deletion, how ever the SLEEP_INTERVAL is to loop on sleeps to wait. It would have been nice if kinesis would have provided out of the box waitTillCreate method

russorat commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,136 @@
+    long waitTime = System.currentTimeMillis() + (10 * 60 * 1000);
+    while (System.currentTimeMillis() < waitTime) {
+      try {
+        Thread.sleep(1000 * 10);
```
but the same sleep time

russorat commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,136 @@
+      try {
+        Thread.sleep(1000 * 10);
+      } catch (InterruptedException e) {
+        LOG.info("Waiting {} milli seconds for the stream to be created", waitTime - System.currentTimeMillis());
```
milliseconds

russorat commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,136 @@
+          return;
+        }
+      } catch (ResourceNotFoundException e) {
+        throw new IllegalStateException(String.format("Stream {} never went active", streamName), e);
```
it didn't go active in the specified wait time.

russorat commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,136 @@
+    try {
+      return kinesisClient.describeStream(describeStreamRequest).getStreamDescription().getStreamStatus();
+    } catch (AmazonServiceException e) {
+      return null;
```
should this be null or empty string? Should you log something here?

russorat commented on core-plugins/widgets/KinesisSink-realtimesink.json:
```diff
@@ -0,0 +1,44 @@
+      "widget-type": "schema",
+      "widget-attributes": {
+        "schema-types": [
+          "string"
```
can this also be bytes?

russorat commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,177 @@
+          numRecordsWritten += writeString(putRecordRequest, data);
+          break;
+        default:
+          LOG.debug("Type {} is not supported for writing to stream", data.getClass().getName());
```
Kinesis only supports bytes and strings? Does that mean before each Kinesis sync, you need to use a projection plugin to convert any other types? Should you add the conversion for the primitive types to string here?

russorat commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,177 @@
+
+  private int writeString(PutRecordRequest putRecordRequest, Object data) {
+    putRecordRequest.setData(ByteBuffer.wrap(Bytes.toBytes((String) data)));
+    kinesisClient.putRecord(putRecordRequest);
```
what happens if this fails? it will still return a 1 even though no record was written.

russorat commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,177 @@
+      return 0;
+    }
+    putRecordRequest.setData(buffer);
+    kinesisClient.putRecord(putRecordRequest);
```
same as above

russorat commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,177 @@
+      buffer = ByteBuffer.wrap((byte []) data);
+    } else {
+      LOG.debug("Type {} is not supported for writing to stream", data.getClass().getName());
+      return 0;
```
should this throw an error?

abhinavbansal246 replied:
I think no, because in realtime there would be multiple requests, failing to write 1 record should not break the pipeline.

russorat commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,177 @@
+    @Description("Partition key to identify shard")
+    private String partition;
+
+
```
extra newline

russorat commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,177 @@
+        continue;
+      }
+      Schema.Field dataSchemaField = schema.getField(config.bodyField);
+      PutRecordRequest putRecordRequest = new PutRecordRequest();
```
This doesn't seem to be doing any batching. Can you leverage multiple records per stream? http://docs.aws.amazon.com/streams/latest/dev/kinesis-kpl-concepts.html#d0e5122

bdmogal replied:
It doesn't seem like this has been addressed?

russorat commented on core-plugins/widgets/KinesisSink-realtimesink.json:
```diff
@@ -0,0 +1,44 @@
+        {
+          "widget-type": "textbox",
+          "label": "Reference Name",
+          "name": "name"
```
no bodyField?

russorat commented on core-plugins/widgets/KinesisSink-realtimesink.json:
```diff
@@ -0,0 +1,44 @@
+        {
+          "widget-type": "textbox",
+          "label": "Partition Key",
+          "name": "partitionKey"
```
no shardCount?

russorat commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,177 @@
+      this.awsAccessSecret = awsAccessSecret;
+      this.partition = partition;
+      this.shardCount = shardCount;
+    }
```
Do any of these properties need to be validated?

abhinavbansal246 replied:
creating AWS client will throw error if the inputs are invalid

russorat commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,177 @@
+                                "Access Key secret required");
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(config.partition),
+                                "partition name should be non-null, non-empty.");
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(config.bodyField),
```
Isn't this the inverse of what it should be checking? ! isNullOrEmpty = false for null or empty values?

abhinavbansal246 replied:
checkArgument will pass if the condition passed in it is True

russorat commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,177 @@
+@Description("Real-time sink that outputs to a specified AWS Kinesis stream.")
+public class RealtimeKinesisStreamSink extends RealtimeSink<StructuredRecord> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RealtimeSink.class);
```
update logger class

russorat commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,177 @@
+          LOG.debug("Type {} is not supported for writing to stream", data.getClass().getName());
+          break;
+      }
+
```
extra newline

russorat commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,177 @@
+    ByteBuffer buffer;
+    if (data instanceof ByteBuffer) {
+      buffer = (ByteBuffer) data;
+    } else if (data instanceof byte []) {
```
does this pass checkstyle with the space between byte and []?

russorat commented:
missing an updated description.md file

anwar6953 commented on core-plugins/pom.xml:
```diff
@@ -173,6 +173,12 @@
       <version>0.6.3</version>
     </dependency>
+    <!-- https://mvnrepository.com/artifact/com.amazonaws/amazon-kinesis-connectors -->
+    <dependency>
```
how much larger does this dependency make the core-plugins JAR?
How large is the amazon-kinesis-connectors jar(s)?

abhinavbansal246 replied:
Its a small jar of 52kb

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,138 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(KinesisUtil.class);
+  private static final long DELETE_TIMEOUT = 1000 * 120;
+  private static final long CREATE_TIMEOUT = 10 * 60 * 1000;
```
10 minute timeout for create? That seems excessive, no?
Is amazon Kinesis that slow?

abhinavbansal246 replied:
CREATE_TIMEOUT is the maximum time to wait and a retry is made after every sleep_interval, originally assigned it 10 minutes to accommodate network or outage issues in Kinesis. Changing it to 3 mins

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,138 @@
+  private static final long DELETE_TIMEOUT = 1000 * 120;
+  private static final long CREATE_TIMEOUT = 10 * 60 * 1000;
+  private static final long SLEEP_INTERVAL = 1000 * 10;
+  /**
```
put a space before this line

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,138 @@
+   * @param streamName The Amazon Kinesis stream name to create
+   * @param shardCount The shard count to create the stream with
+   * @throws IllegalStateException Invalid Amazon Kinesis stream state
+   * @throws IllegalStateException Stream does not go active before the timeout
```
go -> become

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,138 @@
+            try {
+              LOG.info("...Deleting Stream {} ...", streamName);
+              Thread.sleep(SLEEP_INTERVAL);
+            } catch (InterruptedException e) {
```
I don't think its a good idea to just catch InterruptedException.
That might not allow the user to stop the workflow if its running in this part of the code, then?

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,138 @@
+      String state = streamState(kinesisClient, streamName);
+      switch (state) {
+        case "DELETING":
+          long waitTimeDelete = System.currentTimeMillis() + DELETE_TIMEOUT;
```
can you explain why we want to wait for the stream to finish deletion before we throw an exception?

abhinavbansal246 replied:
@anwar6953 in case the stream is being deleted, If it gets deleted within DELETE_TIMEOUT , new stream with that name is created, if it does not we throw an exception

anwar6953 replied:
in the case its being deleted, where is the new stream with that name being created?

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,138 @@
+          LOG.info("Stream {} is UPDATING", streamName);
+          return;
+        default:
+          throw new IllegalStateException("Illegal stream state: " + state);
```
`streamState` method can return empty string. When it gets here, its not a very useful error message.

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,138 @@
+      createStreamRequest.setStreamName(streamName);
+      createStreamRequest.setShardCount(shardCount);
+      kinesisClient.createStream(createStreamRequest);
+      LOG.info("Stream {} created", streamName);
```
nit - also include the shardCount in the log message?

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,138 @@
+      kinesisClient.createStream(createStreamRequest);
+      LOG.info("Stream {} created", streamName);
+    }
+    long waitTimeCreate = System.currentTimeMillis() + CREATE_TIMEOUT;
```
If the stream remains in `CREATING` for more than the `CREATE_TIMEOUT`, it doesn't seem like your code handles that case properly.

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/Properties.java:
```diff
@@ -156,6 +156,18 @@
+   */
+  public static class KinesisRealtimeSink {
+    public static final String NAME = "name";
+    public static final String BODY_FIELD = "body.field";
```
why the inconsistency - `body.field` vs `partitionKey`

abhinavbansal246 replied:
this is something I observed in the properties of other plugins. We tend to have dot.separated for cdap specific properties like "dataset.cube.properties" in cube and camelCase for external config properties eg basePath in TPFS

bdmogal replied:
basePath is not external, since TPFS is very much a CDAP dataset too. I agree there should be standardization, but it seems like we use camel case for everything in hydrator.

bdmogal replied:
let's change this to `bodyField`?

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/Properties.java:
```diff
@@ -156,6 +156,18 @@
+   */
+  public static class KinesisRealtimeSink {
+    public static final String NAME = "name";
+    public static final String BODY_FIELD = "body.field";
```
dot.separated vs camelCase

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,199 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(RealtimeKinesisStreamSink.class);
+  private final KinesisConfig config;
+  private static AmazonKinesisClient kinesisClient;
```
why does this need to be static?

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,199 @@
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(config.name),
+                                "Stream name should be non-null, non-empty.");
```
why inconsistent error messages here (compared to the messages below it)

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,199 @@
+      return 1;
+    } catch (Exception e) {
+      LOG.debug("could not write data to stream {}", putRecordRequest.getStreamName(), e);
+      e.printStackTrace();
```
remove the `printStackTrace` call. The stacktrace can be logged by passing `e` as the last parameter to a `LOG` method.

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,199 @@
+    int numRecordsWritten = 0;
+    for (StructuredRecord structuredRecord : structuredRecords) {
+      Schema schema = structuredRecord.getSchema();
+      Object data = structuredRecord.get(config.bodyField);
```
see if you can do some validation (at configure time) that that input has a field named `config.bodyField`

anwar6953 commented on core-plugins/docs/KinesisSink-realtimesink.md:
```diff
@@ -0,0 +1,48 @@
+**name:** The name of the kinesis stream to output to. Must be a valid kinesis stream name. The kinesis stream
+will be created if it does not exist.
+
+**body.field:** Name of the field in the record that contains the data to be written to
```
Based upon the code, it looks like that its also allowed to be an int or long or other simple types.

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,199 @@
+    return numRecordsWritten;
+  }
+
+  private int writeString(PutRecordRequest putRecordRequest, Object data) {
```
I think this method should return a boolean, instead of 1 or 0.

abhinavbansal246 replied:
@override write method inherited from realtime sink, is expected to return the number of records written successfully. These helper write methods are returning ints to increment the numRecordsWritten in write method.

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,199 @@
+    putRecordRequest.setData(ByteBuffer.wrap(Bytes.toBytes((String) data)));
+    try {
+      PutRecordResult result = kinesisClient.putRecord(putRecordRequest);
+      LOG.info("Data written at {} ", result.toString());
```
no need to call `toString()`

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,199 @@
+      LOG.info("Data written at {} ", result.toString());
+      return 1;
+    } catch (Exception e) {
+      LOG.debug("could not write data to stream {}", putRecordRequest.getStreamName(), e);
```
I don't know if this is proper behavior.
Either we should propagate the exception, or log it at `WARN`.
@albertshau , suggestions?

abhinavbansal246 replied:
due to the realtime nature of the sink, if we propagate error the subsequent data will also be not written to kinesis. changed to warning

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,199 @@
+    } else if (data instanceof byte[]) {
+      buffer = ByteBuffer.wrap((byte[]) data);
+    } else {
+      LOG.debug("Type {} is not supported for writing to stream", data.getClass().getName());
```
same here. this should at least be `WARN`

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,199 @@
+    putRecordRequest.setData(buffer);
+    try {
+      PutRecordResult result = kinesisClient.putRecord(putRecordRequest);
+      LOG.info("Data written at {} ", result.toString());
```
Don't `LOG.info` for every record written.

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,199 @@
+      LOG.info("Data written at {} ", result.toString());
+      return 1;
+    } catch (Exception e) {
+      LOG.debug("could not write data to stream {}", putRecordRequest.getStreamName(), e);
```
propagate the exception or make it a `WARN`.
Ask others for input on good practice in this situation.

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,199 @@
+    }
+  }
+
+  public KinesisConfig getConfig() {
```
Do other plugins expose their configs via `getConfig`? Seems unused

anwar6953 commented on core-plugins/src/test/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSinkTest.java:
```diff
@@ -0,0 +1,76 @@
+
+
+  @Test(expected = IllegalStateException.class)
+  public void kinesistest() throws Exception {
```
We can't test the positive case in unit tests?
You've been testing this against cluster?

abhinavbansal246 replied:
kinesis is configured using api access keys, I have a unit test for checking positive test cases that i can not check in because of the presence of api keys in it.
tested on cluster

bdmogal replied:
doesn't kinesis provide a testing utility? Perhaps we should figure out a way to write an integration test then.

Also, rename the method to indicate what the test does.

bdmogal commented on core-plugins/docs/KinesisSink-realtimesink.md:
```diff
@@ -0,0 +1,47 @@
+
+Example
+-------
+This example will write to a kinesis stream named 'MyKinesisStream'. The kinesis stream will be created if it does not exists already.
```
trim lines to 120 characters

bdmogal commented on core-plugins/docs/KinesisSink-realtimesink.md:
```diff
@@ -0,0 +1,47 @@
+-------
+This example will write to a kinesis stream named 'MyKinesisStream'. The kinesis stream will be created if it does not exists already.
+Each record it receives will be written as a single stream event. The stream event body will be equal to the value of the 'message' field
+from the input record.
```
a line break is not needed before `from`. It should be before either `value` or `of`.

bdmogal replied:
also add `::` to the end of the line instead of a `.`

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,150 @@
+   * @throws IllegalStateException Invalid Amazon Kinesis stream state
+   * @throws IllegalStateException Stream does not become active before the timeout
+   */
+  public static void createAndWaitForStreamToBecomeAvailable(AmazonKinesisClient kinesisClient,
```
this method is only used in the sink. Move it there? Shouldn't create public static methods unless absolutely necessary. Also, can you make the method name a little shorter?

abhinavbansal246 replied:
we are creating a kinesis sink and also Batch plugins for kinesis. We will be using these methods for all of these

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,150 @@
+          long waitTimeDelete = System.currentTimeMillis() + DELETE_TIMEOUT;
+          while (System.currentTimeMillis() < waitTimeDelete && streamExists(kinesisClient, streamName)) {
+            try {
+              LOG.info("...Deleting Stream {} ...", streamName);
```
debug

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,150 @@
+/**
+ * Utility methods to help create and manage Kinesis Streams
+ */
+public class KinesisUtil {
```
it doesn't seem like this class is necessary, since only the KinesisSink uses it. 

If it is necessary, make it a final class with a private constructor.

abhinavbansal246 replied:
source will be using it. made class final

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,150 @@
+   * @param streamName The Amazon Kinesis stream to get the state of
+   * @return String representation of the Stream state
+   */
+  private static String streamState(AmazonKinesisClient kinesisClient, String streamName) {
```
`getStreamState`

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,150 @@
+    if (streamExists(kinesisClient, streamName)) {
+      String state = streamState(kinesisClient, streamName);
+      switch (state) {
+        case "DELETING":
```
aren't there constants for these?

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,213 @@
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(config.awsAccessSecret),
+                                "Access Key secret should be non-null, non-empty.");
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(config.partition),
+                                "partition name should be non-null, non-empty.");
```
for consistency, upper case P

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,213 @@
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(config.partition),
+                                "partition name should be non-null, non-empty.");
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(config.bodyField),
+                                "input field should be non-null, non-empty.");
```
ditto

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,213 @@
+      Schema.Type fieldType = fieldSchema.isNullable() ? fieldSchema.getNonNullable().getType() : fieldSchema.getType();
+      if (!fieldType.isSimpleType()) {
+        throw new IllegalArgumentException(
+          "Type for field  " + config.bodyField + " must be primitive type");
```
Should mention what types are supported, and what was found instead.

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,213 @@
+    BasicAWSCredentials awsCred = new BasicAWSCredentials(config.awsAccessKey, config.awsAccessSecret);
+    kinesisClient = new AmazonKinesisClient(awsCred);
+    createAndWaitForStreamToBecomeAvailable(kinesisClient,
+                                            config.name,
```
nit: no need for newlines here, can merge this into 1 line

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,213 @@
+    putRecordRequest.setData(ByteBuffer.wrap(Bytes.toBytes((String) data)));
+    try {
+      PutRecordResult result = kinesisClient.putRecord(putRecordRequest);
+      LOG.info("Data written at {} ", result);
```
shouldn't log info on every record

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,213 @@
+      buffer = (ByteBuffer) data;
+    } else if (data instanceof byte[]) {
+      buffer = ByteBuffer.wrap((byte[]) data);
+    } else {
```
this should never happen right? this method is always passed a byte array or a byte buffer? Just throw an IllegalStateException from the else, otherwise it seems like this is a valid scenario while reading the code.

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,213 @@
+    private String name;
+
+    @Name(Properties.KinesisRealtimeSink.BODY_FIELD)
+    @Description("Name of the field in the record that contains the data to be " +
```
nit: make the line breaks consistent around 120 characters?

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,213 @@
+
+    @Name(Properties.KinesisRealtimeSink.ACCESS_ID)
+    @Description("AWS access Id having access to Kinesis streams")
+    private String awsAccessKey;
```
all these can be nullable, right?

abhinavbansal246 replied:
all these are required, I am checking that in configure pipeline

bdmogal replied:
oh oops, yes, I meant final?

bdmogal commented on core-plugins/src/test/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSinkTest.java:
```diff
@@ -0,0 +1,76 @@
+    Id.Application appId = Id.Application.from(Id.Namespace.DEFAULT, "testCubeSink");
+    AppRequest<ETLRealtimeConfig> appRequest = new AppRequest<>(APP_ARTIFACT, etlConfig);
+    deployApplication(appId, appRequest);
+
```
nit: remove empty line

bdmogal commented on core-plugins/src/test/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSinkTest.java:
```diff
@@ -0,0 +1,76 @@
+    deployApplication(appId, appRequest);
+
+  }
+
```
ditto

bdmogal commented on core-plugins/src/test/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSinkTest.java:
```diff
@@ -0,0 +1,76 @@
+  }
+
+}
+
```
ditto: here and next line

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,150 @@
+          return;
+        }
+      } catch (ResourceNotFoundException e) {
+        throw new IllegalStateException(String.format("Stream %s did not go active in %d",
```
become active

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,150 @@
+      }
+    }
+    if (!streamState(kinesisClient, streamName).equals("ACTIVE")) {
+      throw new IllegalStateException(String.format("Stream %s did not go active in %d",
```
become active

anwar6953 commented on core-plugins/docs/KinesisSink-realtimesink.md:
```diff
@@ -17,8 +17,7 @@ Properties
 **body.field:** Name of the field in the record that contains the data to be written to
-the specified kinesis stream. The data could be in binary format as a byte array or a ByteBuffer.
-It can also be a String.
+the specified kinesis stream. The data could be in binary format as a byte array , a ByteBuffer or it can be any primitive type.
```
nit, remove the space before the comma

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -31,16 +31,17 @@
   private static final long SLEEP_INTERVAL = 1000 * 10;
-  /**
+
+   /**
```
nit - dedent the `/**`

anwar6953 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -64,14 +63,34 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    validateInputSchema(inputSchema);
+  }
+
+  void validateInputSchema(Schema inputSchema) {
```
make the method private.
Mark the parameter `@Nullable`.

anwar6953 commented on core-plugins/src/test/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSinkTest.java:
```diff
@@ -68,9 +68,5 @@ public void kinesistest() throws Exception {
-}
-
-
+}
```
put a newline at end of file

anwar6953 commented:
How do you test this?
Is the entire setup and testing process documented somewhere? Can we add an integration test for it?

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -88,29 +97,29 @@ public static void createAndWaitForStreamToBecomeAvailable(AmazonKinesisClient k
-        String streamStatus = streamState(kinesisClient, streamName);
-        if (streamStatus.equals("ACTIVE")) {
+        String streamStatus = getStreamState(kinesisClient, streamName);
+        if (streamStatus.equals(STATUS_ACTIVE)) {
```
STATUS_ACTIVE.equals(streamStatus)

gokulavasan committed:
# This is a combination of 2 commits.
# The first commit's message is:

Merge pull request #441 from caskdata/feature/remove-proto-dep

HYDRATOR-913 Remove cdap-proto dependency from Hydrator Plugins

# This is the 2nd commit message:

Adding Aws Kinesis Realtime sink

bdmogal commented on core-plugins/docs/KinesisSink-realtimesink.md:
```diff
@@ -0,0 +1,47 @@
+**name:** The name of the kinesis stream to output to. Must be a valid kinesis stream name. The kinesis stream
+will be created if it does not exist.
+
+**body.field:** Name of the field in the record that contains the data to be written to the specified kinesis stream.
```
this should be changed to `bodyField`

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,159 @@
+   * @throws IllegalStateException Stream does not become active before the timeout
+   */
+  public static void createAndWaitForStream(AmazonKinesisClient kinesisClient,
+                                            String streamName,
```
nit: these line breaks may not be necessary

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,159 @@
+          long waitTimeDelete = System.currentTimeMillis() + DELETE_TIMEOUT;
+          while (System.currentTimeMillis() < waitTimeDelete && streamExists(kinesisClient, streamName)) {
+            try {
+              LOG.debug("...Deleting Stream {} ...", streamName);
```
nit: remove the leading `...`

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,159 @@
+public final class KinesisUtil {
+
+  private static final Logger LOG = LoggerFactory.getLogger(KinesisUtil.class);
+  private static final long DELETE_TIMEOUT = 1000 * 120;
```
why is the delete timeout value different from create timeout? Are they known to take different times? I'd suggest making them the same, and turning them into nullable plugin configurations with default values

bdmogal replied:
If you turn them into configurations, also validate that SLEEP_INTERVAL < TIMEOUT

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,159 @@
+  private static final long CREATE_TIMEOUT = 3 * 60 * 1000;
+  private static final long SLEEP_INTERVAL = 1000 * 10;
+
+  private static final String STATUS_CREATING = "CREATING";
```
I didn't mean earlier for you to define these constants. I thought the Kinesis client would provide them?

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,159 @@
+            }
+          }
+          if (streamExists(kinesisClient, streamName)) {
+            throw new IllegalStateException(String.format("KinesisUtils timed out waiting for stream {} to delete",
```
nit: remove `KinesisUtils` from the message

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,159 @@
+      String state = getStreamState(kinesisClient, streamName);
+      switch (state) {
+        case STATUS_DELETING:
+          long waitTimeDelete = System.currentTimeMillis() + DELETE_TIMEOUT;
```
Instead of calling `System.currentTimeMillis()` twice, which could return two different values, extract it to a variable `deleteStartTime` and use it?

abhinavbansal246 replied:
the second System.currentTimeMillis() is in the check condition of the while loop, it fetches the current timestamp at every check (after sleep_interval). putting it in a variable will result in an infinite loop as deleteStartTime will never be > waitTimeDelete

bdmogal replied:
right, I meant in the loop, you would reset currentTime. This was a little nitpicky, though.

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,159 @@
+          while (System.currentTimeMillis() < waitTimeDelete && streamExists(kinesisClient, streamName)) {
+            try {
+              LOG.debug("...Deleting Stream {} ...", streamName);
+              Thread.sleep(SLEEP_INTERVAL);
```
For consistency, use `TimeUnit.MILLISECONDS.sleep(SLEEP_INTERVAL)`

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,159 @@
+          createStream(streamName, shardCount, kinesisClient);
+          break;
+        case STATUS_ACTIVE:
+          LOG.info("Stream {} is ACTIVE", streamName);
```
since ACTIVE/UPDATING are Kinesis-internal statuses,  better to say `Stream {} already exists` here and `Stream {} is being updated` on line 84?

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,159 @@
+    } else {
+      createStream(streamName, shardCount, kinesisClient);
+    }
+    long waitTimeCreate = System.currentTimeMillis() + CREATE_TIMEOUT;
```
same comment about not calling `System.currentTimeMillis()` twice.

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,159 @@
+      createStream(streamName, shardCount, kinesisClient);
+    }
+    long waitTimeCreate = System.currentTimeMillis() + CREATE_TIMEOUT;
+    while (System.currentTimeMillis() < waitTimeCreate) {
```
Any way that this wait loop can be extracted to a method and used from both here and line 63? Something like - 

```
private void waitForStreamState(AmazonKinesisClient kinesisClient, String streamName, String expectedStreamState, long timeout, long sleepInterval) throws TimeoutException, InterruptedException {
  long currentTime = System.currentTimeMillis();
  long waitUntilTime = currentTime + timeout;
  while (currentTime < waitUntilTime) {
    TimeUnit.MILLISECONDS.sleep(SLEEP_INTERVAL);
    String currentStreamState = getStreamState(kinesisClient, streamName);
    if (expectedStreamState.equals(currentStreamState) {
      return;
    }
  }
  if (!expectedStreamState.equals(getStreamState(kinesisClient, streamName))) {
    // throw Exception
  }
}
```

abhinavbansal246 replied:
Both the loops have different check conditions (StreamExists while deleting and just timeout check for creating. moving them into a single method might result in hard to read complex code for a simple task

bdmogal replied:
not sure what you mean. Both of them should be a streamExists check right? For delete, you bail if stream exists after timeout, and for create, you bail if stream does not exist after timeout?

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,159 @@
+    createStreamRequest.setStreamName(streamName);
+    createStreamRequest.setShardCount(shardCount);
+    kinesisClient.createStream(createStreamRequest);
+    LOG.info("Stream " + streamName + " created");
```
use string format syntax from Logger instead of concatenation

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,159 @@
+    try {
+      return kinesisClient.describeStream(describeStreamRequest).getStreamDescription().getStreamStatus();
+    } catch (AmazonServiceException e) {
+      LOG.debug("State of the stream {} could not be found", streamName);
```
include the exception in the debug message

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,159 @@
+      return kinesisClient.describeStream(describeStreamRequest).getStreamDescription().getStreamStatus();
+    } catch (AmazonServiceException e) {
+      LOG.debug("State of the stream {} could not be found", streamName);
+      return "";
```
hmmmm, not a good idea to return empty string. just propagate the exception from this method?

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,159 @@
+   * @return true if the Amazon Kinesis stream exists, otherwise return false
+   */
+  private static boolean streamExists(AmazonKinesisClient kinesisClient, String streamName) {
+    DescribeStreamRequest describeStreamRequest = new DescribeStreamRequest();
```
instead of doing this all over again, can you refactor to just call the above method from here and do appropriate exception handling?

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,210 @@
+  }
+
+  private void validateInputSchema(@Nullable Schema inputSchema) {
+    if (inputSchema != null) {
```
what happens when `inputSchema` is null?

Also, to reduce indentation, we prefer:

```
if (inputSchema == null) {
  return;
}
// Rest of the logic ...
```

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,210 @@
+      Schema.Type fieldType = fieldSchema.isNullable() ? fieldSchema.getNonNullable().getType() : fieldSchema.getType();
+      if (!fieldType.isSimpleType()) {
+        throw new IllegalArgumentException(String.format(
+          "Type %s for field %s is not supported. Supported types are Int, long, float, string, Bytes", fieldType,
```
nit: should standardize on using types from `Schema.Type`, since those are the ones that users will set?

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,210 @@
+
+  @Override
+  public void initialize(RealtimeContext context) throws Exception {
+    super.initialize(context);
```
I am not sure, but do you need this?

bdmogal replied:
could you check this?

abhinavbansal246 replied:
Yes sir, we do

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,210 @@
+  public void initialize(RealtimeContext context) throws Exception {
+    super.initialize(context);
+    BasicAWSCredentials awsCred = new BasicAWSCredentials(config.awsAccessKey, config.awsAccessSecret);
+    kinesisClient = new AmazonKinesisClient(awsCred);
```
It looks like `AmazonKinesisClient` has a `shutdown` method that should be called in the `destroy` method of the plugin - http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/AmazonWebServiceClient.html#shutdown--

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,210 @@
+import java.nio.ByteBuffer;
+import javax.annotation.Nullable;
+
+import static co.cask.hydrator.plugin.common.KinesisUtil.createAndWaitForStream;
```
nit: We generally don't use static imports. Instead qualify the usages with the class name

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,210 @@
+          numRecordsWritten += writeString(putRecordRequest, data);
+          break;
+        default:
+          if (dataSchemaField.getSchema().getType().isSimpleType()) {
```
since you're already `switch`ing on `dataSchemaField.getSchema().getType()`, just call the supported ones out in `switch` cases, and only use the `default` case for the unsupported types.

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,210 @@
+                                                    data.getClass().getName()));
+    }
+    putRecordRequest.setData(buffer);
+    try {
```
extract lines 160-166 in a separate method and use on line 140 and here.

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,210 @@
+   */
+  public static class KinesisConfig extends PluginConfig {
+
+    @Description("The name of the Kinesis stream to output to. Must be a valid Kinesis stream name. The Kinesis " +
```
should mention what invalid stream names are?

bdmogal commented on core-plugins/src/test/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSinkTest.java:
```diff
@@ -0,0 +1,72 @@
+    AppRequest<ETLRealtimeConfig> appRequest = new AppRequest<>(APP_ARTIFACT, etlConfig);
+    deployApplication(appId, appRequest);
+  }
+}
```
add newline at the end of the file. Does the build not fail?

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -24,22 +24,17 @@
   private static final Logger LOG = LoggerFactory.getLogger(KinesisUtil.class);
-  private static final long DELETE_TIMEOUT = 1000 * 120;
-  private static final long CREATE_TIMEOUT = 3 * 60 * 1000;
+  private static final long TIMEOUT = 3 * 60 * 1000;
```
so we're not making these configurable?

abhinavbansal246 replied:
Making it configurable might be an overkill since it is needed for creation of stream and has nothing to do with data writing operations. If you suggest we can make it configurable

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -52,64 +47,61 @@ private KinesisUtil() {
       switch (state) {
-        case STATUS_DELETING:
-          long waitTimeDelete = System.currentTimeMillis() + DELETE_TIMEOUT;
+        case "DELETING":
```
hmmm, so there's no constants in Kinesis client for these?

abhinavbansal246 replied:
nothing that Kinesis sdk has exposed

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,148 @@
+    if (streamExists(kinesisClient, streamName)) {
+      String state = getStreamState(kinesisClient, streamName);
+      switch (state) {
+        case "DELETING":
```
You should use this enum - http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/kinesis/model/StreamStatus.html

chtyim commented:
Seems like you did a wrong merge. Now this PR touches so many files and the version get changed to non-SNAPSHOT version

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,150 @@
+  private static final Logger LOG = LoggerFactory.getLogger(KinesisUtil.class);
+  private static final long TIMEOUT = 3 * 60 * 1000;
+  private static final long SLEEP_INTERVAL = 1000 * 10;
+  private KinesisUtil() {
```
add a newline before this

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,150 @@
+      createStream(streamName, shardCount, kinesisClient);
+    }
+    long waitTimeCreate = System.currentTimeMillis() + TIMEOUT;
+    while (System.currentTimeMillis() < waitTimeCreate) {
```
I still don't see why this code cannot be refactored into a method.

bdmogal replied:
both create and delete eventually call `getStreamState` as the exit condition from the loop.

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,150 @@
+    createStreamRequest.setStreamName(streamName);
+    createStreamRequest.setShardCount(shardCount);
+    kinesisClient.createStream(createStreamRequest);
+    LOG.info("Stream {} is being created", streamName);
```
nit: it already got created by this time

abhinavbansal246 replied:
createStream method just initializes the creation. Actual creation happens after a while

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -0,0 +1,150 @@
+    try {
+      String status = kinesisClient.describeStream(describeStreamRequest).getStreamDescription().getStreamStatus();
+      StreamStatus streamStatus = StreamStatus.fromValue(status);
+     return streamStatus;
```
nit: indentation

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSink.java:
```diff
@@ -0,0 +1,217 @@
+    Schema.Type fieldType = fieldSchema.isNullable() ? fieldSchema.getNonNullable().getType() : fieldSchema.getType();
+    if (!fieldType.isSimpleType()) {
+      throw new IllegalArgumentException(String.format(
+        "Type %s for field %s is not supported. Supported types are Boolean, INT, LONG, FLOAT, DOUBLE, STRING, Bytes",
```
why are Boolean and Bytes not all caps? Infact, why not just use the enum itself?

bdmogal commented on core-plugins/src/test/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSinkTest.java:
```diff
@@ -0,0 +1,72 @@
+ */
+public class RealtimeKinesisStreamSinkTest extends ETLRealtimeTestBase {
+
+
```
nit: extra empty line

bdmogal commented on core-plugins/src/test/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSinkTest.java:
```diff
@@ -0,0 +1,72 @@
+import java.util.Map;
+
+/**
+ * Test default configuration
```
doesn't seem like an appropriate description. Just "Tests for {@link RealtimeKinesisStreamSink}"?

bdmogal commented on core-plugins/src/test/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSinkTest.java:
```diff
@@ -0,0 +1,72 @@
+                                                     Constants.Reference.REFERENCE_NAME, "DG"),
+                                     null);
+
+
```
nit: extra empty line

bdmogal commented on core-plugins/src/test/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSinkTest.java:
```diff
@@ -0,0 +1,72 @@
+    properties.put(Properties.KinesisRealtimeSink.PARTITION_KEY, "part");
+
+    ETLPlugin sink = new ETLPlugin("KinesisSink", RealtimeSink.PLUGIN_TYPE, properties
+      , null);
```
move to previous line

bdmogal commented on core-plugins/src/test/java/co/cask/hydrator/plugin/realtime/sink/RealtimeKinesisStreamSinkTest.java:
```diff
@@ -0,0 +1,72 @@
+                                     null);
+
+
+    Map<String, String> properties = new HashMap<>();
```
You can use `ImmutableMap.of(....)` instead

abhinavbansal246 replied:
immutable map.of allows a maximum of 5 arguments, the number of properties that needs to be set for kinesis is more

bdmogal commented:
@abhinavbansal246 still concerned about repeat code. Also a few other minor comments.

abhinavbansal-246 committed:
Addressing final review comments for Kinesis stream

abhinavbansal-246 committed:
Addressing  review comments for Kinesis stream

bdmogal commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java:
```diff
@@ -74,32 +71,25 @@ public static void createAndWaitForStream(AmazonKinesisClient kinesisClient, Str
         return;
       }
     }
+    throw new IllegalStateException(String.format("TimedOut waithing for stream to be in %s state", expectedStatus));
```
nit: `Timed out...`

bdmogal commented:
tiny nit. Otherwise LGTM. Please squash and merge after successful build and testing on a cluster.

abhinavbansal-246 committed:
Addressing review comments

abhinavbansal246 commented:
https://github.com/caskdata/hydrator-plugins/pull/451
